src/codecs.rs

```diff
@@ -0,0 +1,120 @@
+use bytes::{BufMut, BytesMut};
+use codec::{Encoder, Decoder};
+use std::{io, str};
+
+/// A simple `Codec` implementation that just ships bytes around.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct Bytes;
+
+/// A simple `Codec` implementation that just ships bytes around.
+// Note: while this constant is technically reduntant above,
+// it's nice to include so that `codec::BYTES` and `codec::LINES`
+// provide the same interface to the end-user.
+pub const BYTES: Bytes = Bytes;
+
+impl Decoder for Bytes {
+    type Item = BytesMut;
+    type Error = io::Error;
+
+    fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<BytesMut>, io::Error> {
+        if buf.len() > 0 {
+            let len = buf.len();
+            Ok(Some(buf.split_to(len)))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<BytesMut>, io::Error> {
+        self.decode(buf)
+    }
+}
+
+impl Encoder for Bytes {
+    type Item = Vec<u8>;
+    type Error = io::Error;
+
+    fn encode(&mut self, data: Vec<u8>, buf: &mut BytesMut) -> Result<(), io::Error> {
+        buf.put(&data[..]);
+        Ok(())
+    }
+}
+
+/// A simple `Codec` implementation that splits up data into lines.
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct Lines {
+    // Stored index of the next index to examine for a `\n` character.
+    // This is used to optimize searching.
+    // For example, if `decode` was called with `abc`, it would hold `3`,
+    // because that is the next index to examine.
+    // The next time `decode` is called with `abcde\n`, the method will
+    // only look at `de\n` before returning.
+    next_index: usize,
+}
+
+/// A simple `Codec` implementation that splits up data into lines.
+pub const LINES: Lines = Lines { next_index: 0 };
```

Comment:
Instead of a `const`, could we perhaps use a `new` constructor? I think these don't necessarily need to be used in a static context, right?

Reply:
(similarly this'd allow `Bytes` above to be an opaque structure and instead just have a `new` constructor as well)

Reply:
@alexcrichton Sure, sounds good to me! Would you prefer `Bytes::new()` or `bytes()`?

Reply:
I'd go with `Bytes::new`

Reply:
Done.